text/0000-macros-1.1.md

```diff
@@ -0,0 +1,552 @@
+a macro context, it uses the passed in [plugin registry][registry] to register
+custom macros, attributes, etc.
+
+[registry]: https://github.com/rust-lang/rust/blob/78d49bfac2bbcd48de522199212a1209f498e834/src/librustc_plugin/registry.rs#L30-L69
+
+After a macro is registered, the compiler will then continue the normal process
+of expanding a crate. Whenever the compiler encounters this macro it will call
+this registration with essentially and AST and morally gets back a different
+AST to splice in or replace.
+
+### Today's drawbacks
+
+This expansion process suffers from many of the downsides mentioned in the
+motivation section, such as a lack of hygiene, a lack of modularity, and the
+inability to import macros as you would normally other functionality in the
+module system.
+
+Additionally, though, it's essentially impossible to ever *stabilize* because
+the interface to the compiler is... the compiler! We clearly want to make
+changes to the compiler over time, so this isn't acceptable. To have a stable
+interface we'll need to cut down this surface area *dramatically* to a curated
+set of known-stable APIs.
+
+Somewhat more subtly, the technical ABI of procedural macros is also exposed
+quite thinly today as well. The implementation detail of dynamic libraries, and
+especially that both the compiler and the macro dynamically link to libraries
+like libsyntax, cannot be changed. This precludes, for example, a completely
+statically linked compiler (e.g. compiled for `x86_64-unknown-linux-musl`).
+Another goal of this RFC will also be to hide as many of these technical
+details as possible, allowing the compiler to flexibly change how it interfaces
+to macros.
+
+## Macros 1.1
+
+Ok, with the background knowledge of what procedural macros are today, let's
+take a look at how we can solve the major problems blocking its stabilization:
+
+* Sharing an API of the entire compiler
+* Frozen interface between the compiler and macros
+
+### `librustc_macro`
+
+Proposed in [RFC 1566](https://github.com/rust-lang/rfcs/pull/1566) and
+described in [this blog post](http://ncameron.org/blog/libmacro/) the
+distribution will now ship with a new `librustc_macro` crate available for macro
+authors. The intention here is that the gory details of how macros *actually*
+talk to the compiler is entirely contained within this one crate. The stable
+interface to the compiler is then entirely defined in this crate, and we can
+make it as small or large as we want. Additionally, like the standard library,
+it can contain unstable APIs to test out new pieces of functionality over time.
+
+The initial implementation of `librustc_macro` is proposed to be *incredibly*
+bare bones:
+
+```rust
+#![crate_name = "macro"]
+
+pub struct TokenStream {
+    // ...
+}
+
+#[derive(Debug)]
+pub struct LexError {
+    // ...
+}
+
+pub struct Context {
+    // ...
+}
+
+impl TokenStream {
+    pub fn from_source(cx: &mut Context,
```

Comment:
@nrc could you comment on the `Context` argument here? I believe both @eddyb and @cgswords think it may not be necessary, in which case the pointer could be dropped everywhere and these methods could just be `fmt::Display` and `FromStr` (standard traits)